File: geo/src/test/java/org/odk/collect/geo/support/FakeMapFragment.java
package org.odk.collect.geo.support;

import static java.util.Collections.emptyList;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentManager;

import org.odk.collect.geo.maps.MapFragment;
import org.odk.collect.geo.maps.MapPoint;

import java.util.ArrayList;
import java.util.List;

import kotlin.Pair;

public class FakeMapFragment implements MapFragment {

    private PointListener gpsLocationListener;
    private String locationProvider;
    private boolean retainMockAccuracy;
    private MapPoint center;
    private double zoom;
    private Pair<Iterable<MapPoint>, Float> zoomBoundingBox;
    private MapPoint gpsLocation;
    private FeatureListener featureClickListener;
    private final List<MapPoint> markers = new ArrayList<>();

    public void applyConfig(Bundle config) {

    }

    @Override
    public void addTo(FragmentManager fragmentManager, int containerId, @Nullable ReadyListener readyListener, @Nullable ErrorListener errorListener) {
        readyListener.onReady(this);
    }

    @NonNull
    @Override
    public MapPoint getCenter() {
        return center;
    }

    @Override
    public double getZoom() {
        return zoom;
    }

    @Override
    public void setCenter(@Nullable MapPoint center, boolean animate) {
        this.center = center;
    }

    @Override
    public void zoomToPoint(@Nullable MapPoint center, boolean animate) {
        this.center = center;
    }

    @Override
    public void zoomToPoint(@Nullable MapPoint center, double zoom, boolean animate) {
        zoomBoundingBox = null;

        this.center = center;
        this.zoom = zoom;
    }

    @Override
    public void zoomToBoundingBox(Iterable<MapPoint> points, double scaleFactor, boolean animate) {
        center = null;
        zoom = 0;

        zoomBoundingBox = new Pair(points, scaleFactor);
    }

    @Override
    public int addMarker(MapPoint point, boolean draggable, String iconAnchor) {
        markers.add(point);
        return markers.size() - 1;
    }

    @Override
    public void setMarkerIcon(int featureId, int drawableId) {

    }

    @Override
    public MapPoint getMarkerPoint(int featureId) {
        return null;
    }

    @Override
    public int addDraggablePoly(@NonNull Iterable<MapPoint> points, boolean closedPolygon) {
        return 0;
    }

    @Override
    public void appendPointToPoly(int featureId, @NonNull MapPoint point) {

    }

    @Override
    public void removePolyLastPoint(int featureId) {

    }

    @NonNull
    @Override
    public List<MapPoint> getPolyPoints(int featureId) {
        return emptyList();
    }

    @Override
    public void removeFeature(int featureId) {

    }

    @Override
    public void clearFeatures() {
        markers.clear();
    }

    @Override
    public void setClickListener(@Nullable PointListener listener) {

    }

    @Override
    public void setLongPressListener(@Nullable PointListener listener) {

    }

    @Override
    public void setFeatureClickListener(@Nullable FeatureListener listener) {
        featureClickListener = listener;
    }

    @Override
    public void setDragEndListener(@Nullable FeatureListener listener) {

    }

    @Override
    public void setGpsLocationEnabled(boolean enabled) {

    }

    @Nullable
    @Override
    public MapPoint getGpsLocation() {
        return gpsLocation;
    }

    @Nullable
    @Override
    public String getLocationProvider() {
        return locationProvider;
    }

    @Override
    public void runOnGpsLocationReady(@NonNull ReadyListener listener) {

    }

    @Override
    public void setGpsLocationListener(@Nullable PointListener listener) {
        this.gpsLocationListener = listener;
    }

    @Override
    public void setRetainMockAccuracy(boolean retainMockAccuracy) {
        this.retainMockAccuracy = retainMockAccuracy;
    }

    public void setLocation(MapPoint mapPoint) {
        gpsLocation = mapPoint;

        if (gpsLocationListener != null) {
            gpsLocationListener.onPoint(mapPoint);
        }
    }

    public void setLocationProvider(String locationProvider) {
        this.locationProvider = locationProvider;
    }

    public boolean isRetainMockAccuracy() {
        return retainMockAccuracy;
    }

    public Pair<Iterable<MapPoint>, Float> getZoomBoundingBox() {
        return zoomBoundingBox;
    }

    public void clickOnFeature(int index) {
        featureClickListener.onFeature(index);
    }

    public List<MapPoint> getMarkers() {
        return markers;
    }
}

File: geo/src/test/java/org/odk/collect/geo/support/FakeMapFragment.kt
package org.odk.collect.geo.support

import androidx.fragment.app.Fragment
import org.odk.collect.maps.LineDescription
import org.odk.collect.maps.MapFragment
import org.odk.collect.maps.MapFragment.FeatureListener
import org.odk.collect.maps.MapFragment.PointListener
import org.odk.collect.maps.MapFragment.ReadyListener
import org.odk.collect.maps.MapPoint
import org.odk.collect.maps.PolygonDescription
import org.odk.collect.maps.markers.MarkerDescription
import org.odk.collect.maps.markers.MarkerIconDescription
import kotlin.random.Random

class FakeMapFragment : Fragment(), MapFragment {

    private var clickListener: PointListener? = null
    private var gpsLocationListener: PointListener? = null
    private var locationProvider: String? = null
    private var retainMockAccuracy = false
    private var center: MapPoint? = null
    private var zoom = 0.0
    private var zoomBoundingBox: Pair<Iterable<MapPoint>, Double>? = null
    private var readyListener: ReadyListener? = null
    private var gpsLocation: MapPoint? = null
    private var featureClickListener: FeatureListener? = null
    private val markers = mutableMapOf<Int, MapPoint>()
    private val markerIcons = mutableMapOf<Int, MarkerIconDescription?>()
    private val polyLines = mutableMapOf<Int, LineDescription>()
    private val polygons = mutableMapOf<Int, PolygonDescription>()
    private var hasCenter = false
    private val featureIds = mutableListOf<Int>()

    override fun init(
        readyListener: ReadyListener?,
        errorListener: MapFragment.ErrorListener?
    ) {
        this.readyListener = readyListener
    }

    fun ready() {
        readyListener?.onReady(this)
    }

    override fun getCenter(): MapPoint {
        return center ?: DEFAULT_CENTER
    }

    override fun getZoom(): Double {
        return zoom
    }

    override fun setCenter(center: MapPoint?, animate: Boolean) {
        this.center = center
        hasCenter = true
    }

    override fun zoomToPoint(center: MapPoint?, animate: Boolean) {
        zoomBoundingBox = null
        this.center = center
        this.zoom = DEFAULT_POINT_ZOOM
        hasCenter = true
    }

    override fun zoomToPoint(center: MapPoint?, zoom: Double, animate: Boolean) {
        zoomBoundingBox = null
        this.center = center
        this.zoom = zoom
        hasCenter = true
    }

    override fun zoomToBoundingBox(
        points: Iterable<MapPoint>,
        scaleFactor: Double,
        animate: Boolean
    ) {
        center = null
        zoom = 0.0
        zoomBoundingBox = Pair(
            points.toList(), // Clone list to prevent original changing captured values
            scaleFactor
        )
        hasCenter = true
    }

    override fun addMarker(markerDescription: MarkerDescription): Int {
        val featureId = generateFeatureId()

        markers[featureId] = markerDescription.point
        markerIcons[featureId] = markerDescription.iconDescription

        featureIds.add(featureId)
        return featureId
    }

    override fun addMarkers(markers: List<MarkerDescription>): List<Int> {
        return markers.map {
            addMarker(it)
        }
    }

    override fun setMarkerIcon(featureId: Int, markerIconDescription: MarkerIconDescription) {
        markerIcons[featureId] = markerIconDescription
    }

    override fun getMarkerPoint(featureId: Int): MapPoint {
        return markers[featureId]!!
    }

    override fun addPolyLine(lineDescription: LineDescription): Int {
        val featureId = generateFeatureId()

        polyLines[featureId] = lineDescription
        featureIds.add(featureId)
        return featureId
    }

    override fun addPolygon(polygonDescription: PolygonDescription): Int {
        val featureId = generateFeatureId()
        polygons[featureId] = polygonDescription
        featureIds.add(featureId)
        return featureId
    }

    override fun appendPointToPolyLine(featureId: Int, point: MapPoint) {
        val poly = polyLines[featureId]!!
        polyLines[featureId] = poly.copy(points = poly.points + point)
    }

    override fun removePolyLineLastPoint(featureId: Int) {
        val poly = polyLines[featureId]!!
        polyLines[featureId] = poly.copy(points = poly.points.dropLast(1))
    }

    override fun getPolyLinePoints(featureId: Int): List<MapPoint> {
        return polyLines[featureId]!!.points
    }

    override fun clearFeatures() {
        markers.clear()
        markerIcons.clear()
    }

    override fun setClickListener(listener: PointListener?) {
        this.clickListener = listener
    }

    fun click(point: MapPoint) {
        clickListener?.onPoint(point)
    }

    override fun setLongPressListener(listener: PointListener?) {}
    override fun setFeatureClickListener(listener: FeatureListener?) {
        featureClickListener = listener
    }

    override fun setDragEndListener(listener: FeatureListener?) {}
    override fun setGpsLocationEnabled(enabled: Boolean) {}
    override fun getGpsLocation(): MapPoint? {
        return gpsLocation
    }

    override fun getLocationProvider(): String? {
        return locationProvider
    }

    override fun runOnGpsLocationReady(listener: ReadyListener) {}
    override fun setGpsLocationListener(listener: PointListener?) {
        gpsLocationListener = listener

        gpsLocation?.let {
            listener?.onPoint(it)
        }
    }

    override fun setRetainMockAccuracy(retainMockAccuracy: Boolean) {
        this.retainMockAccuracy = retainMockAccuracy
    }

    override fun hasCenter(): Boolean {
        return hasCenter
    }

    fun setLocation(mapPoint: MapPoint?) {
        gpsLocation = mapPoint
        if (gpsLocationListener != null) {
            gpsLocationListener!!.onPoint(mapPoint!!)
        }
    }

    fun setLocationProvider(locationProvider: String?) {
        this.locationProvider = locationProvider
    }

    fun isRetainMockAccuracy(): Boolean {
        return retainMockAccuracy
    }

    fun clickOnFeature(index: Int) {
        featureClickListener!!.onFeature(featureIds[index])
    }

    fun clickOnFeatureId(featureId: Int) {
        featureClickListener!!.onFeature(featureId)
    }

    fun getMarkers(): List<MapPoint> {
        return markers.values.toList()
    }

    fun getMarkerIcons(): List<MarkerIconDescription?> {
        return markerIcons.values.toList()
    }

    fun getZoomBoundingBox(): Pair<Iterable<MapPoint>, Double>? {
        return zoomBoundingBox
    }

    fun getPolyLines(): List<LineDescription> {
        return polyLines.values.toList()
    }

    fun isPolyClosed(index: Int): Boolean {
        return polyLines[featureIds[index]]!!.closed
    }

    fun isPolyDraggable(index: Int): Boolean {
        return polyLines[featureIds[index]]!!.draggable
    }

    fun getFeatureId(points: List<MapPoint>): Int {
        return if (points.size == 1) {
            markers.entries.find {
                it.value == points[0]
            }!!.key
        } else {
            polyLines.entries.find {
                it.value.points == points
            }!!.key
        }
    }

    private fun generateFeatureId(): Int {
        var featureId = Random.nextInt()
        while (featureIds.contains(featureId)) {
            featureId = Random.nextInt()
        }

        return featureId
    }

    fun getPolygons(): List<PolygonDescription> {
        return polygons.values.toList()
    }

    companion object {
        /**
         * The value returned if the map has had no center set or has had `null` pass to
         * [setCenter]
         */
        val DEFAULT_CENTER = MapPoint(-1.0, -1.0)

        /**
         * The value used to zoom when [zoomToPoint] is called without a zoom level
         */
        const val DEFAULT_POINT_ZOOM = -1.0
    }
}
